scripts/inference/endpoint_generate.py

```diff
@@ -0,0 +1,229 @@
+
+    log.info(f'Using endpoint {url}')
+
+    api_key = os.environ.get(ENDPOINT_API_KEY_ENV, '')
+    if not api_key:
+        log.warning(f'API key not set in {ENDPOINT_API_KEY_ENV}')
+
+    # Load prompts
+    prompt_strings = []
+    for prompt in args.prompts:
+        if prompt.startswith('file::'):
+            prompt = load_prompts_from_file(prompt)
+        prompt_strings.append(prompt)
+
+    cols = ['batch', 'prompt', 'output']
+    param_data = {
+        'max_tokens': args.max_tokens,
+        'temperature': args.temperature,
+        'top_k': args.top_k,
+        'top_p': args.top_p,
+    }
+
+    @sleep_and_retry
+    @limits(calls=args.rate_limit // args.batch_size, period=1)  # type: ignore
+    async def generate(session: aiohttp.ClientSession, batch: int,
+                       prompts: list):
+        data = copy.copy(param_data)
+        data['prompt'] = prompts
+        headers = {'Authorization': api_key, 'Content-Type': 'application/json'}
+
+        req_start = time.time()
+        async with session.post(url, headers=headers, json=data) as resp:
+            if resp.ok:
+                try:
+                    response = await resp.json()
+                except requests.JSONDecodeError:
+                    raise Exception(
+                        f'Bad response: {resp.status_code} {resp.reason}'  # type: ignore
+                    )
+            else:
+                raise Exception(
+                    f'Bad response: {resp.status_code} {resp.content.decode().strip()}'  # type: ignore
+                )
+
+        req_end = time.time()
+        n_compl = response['usage']['completion_tokens']
+        n_prompt = response['usage']['prompt_tokens']
+        req_latency = (req_end - req_start)
+        log.info(f'Completed batch {batch}: {n_compl:,} completion' +
+                 f' tokens using {n_prompt:,} prompt tokens in {req_latency}s')
+
+        res = pd.DataFrame(columns=cols)
+
+        for r in response['choices']:
+            index = r['index']
+            res.loc[len(res)] = [batch, prompts[index], r['text']]
+        return res
+
+    res = pd.DataFrame(columns=cols)
+    batch = 0
+
+    total_batches = math.ceil(len(prompt_strings) / args.batch_size)
+    log.info(
+        f'Generating {len(prompt_strings)} prompts in {total_batches} batches')
+
+    async with aiohttp.ClientSession() as session:
+        tasks = []
+
+        for i in range(total_batches):
+            prompts = prompt_strings[i * args.batch_size:min(
+                (i + 1) * args.batch_size, len(prompt_strings))]
+
+            tasks.append(generate(session, batch, prompts))
+            batch += 1
+
+        results = await asyncio.gather(*tasks)
+        res = pd.concat(results)
+
+    res.reset_index(drop=True, inplace=True)
+    log.info(f'Generated {len(res)} prompts, example data:')
+    log.info(res.head())
+
+    # save res to local output folder
```

Comment:
usually for this kind of thing we just have one `output_path`, and if its local we write locally, if its remote we write remotely. does that work here too?

Reply:
:+1: I can convert `local_folder` into a temp file that we don't parametrize, and then output folder should work as a local dir? Will make that required in this case

Reply:
also, is it actually a folder? or should it just a single file path?

Reply:
It's a folder right now - this seemed more consistent with other scripts.. Should it be a file? Doing `{output_folder}/output.csv` as the final file name

Reply:
If you only have one file in the folder, it should be a file path imo. the others are folder because they actually require folder i think. e.g huggingface checkpoints are folders, not just single files.

Reply:
yea sgtm, 16c9307